### PR TITLE
[ASTMatchers] forCallable should not erase binding on success

### PR DESCRIPTION
**Pull request by marco-antognini-sonarsource**
Do not erase Builder when the first check fails because it could succeed on the second stack frame.

The problem was that `InnerMatcher.matches` erases the bindings when it returns false. The appropriate solution is to pass a copy of the bindings, similar to what `matchesFirstInRange` does.

### CODE REVIEW

**Comment by marco-antognini-sonarsource** [comment]
Could someone review this PR, please? Any feedback you can give me is welcome.

The CI should be green and I've disabled keeping my email private.

Pinging some recent contributors/reviewers to AST Matchers. @danix800, @AaronBallman, @shafik, @PiotrZSL, @steakhal, @cor3ntin

Thanks

**Review by AaronBallman — COMMENTED** [review]
The changes look reasonable to me, but can you also add a release note to clang/docs/ReleaseNotes.rst so that users of AST matchers know about the fix?

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -832,6 +832,7 @@ Bug Fixes to AST Handling
 - Fixed malformed AST generated for anonymous union access in templates. (#GH90842)
 - Improved preservation of qualifiers and sugar in `TemplateNames`, including
   template keyword.
+- ``forCallable`` now properly preserves binding on successful match. (#GH89657)
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:835]: Can you move this under the `AST Matchers` section instead?

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a small nit with the release notes. Do you need someone to land this on your behalf?

File: clang/docs/ReleaseNotes.rst (hunk 1 above)
Comment (marco-antognini-sonarsource) [reply clang/docs/ReleaseNotes.rst]: You're right, I missed this section somehow... it's fixed now.

**Comment by marco-antognini-sonarsource** [comment]
> LGTM aside from a small nit with the release notes. Do you need someone to land this on your behalf?

Thanks for asking. I should be fine -- I'll wait for the CI to complete and merge afterwards.

**Pull request merged**